vispy/app/backends/_tk.py

```diff
@@ -0,0 +1,498 @@
+        # if e.state & 0x0010: mods.append(keys.?)
+        if e.state & 0x0080: mods.append(keys.ALT)
+        # if e.state & 0x0100: mods.append(keys.?)
+        # if e.state & 0x0200: mods.append(keys.?)
+        # if e.state & 0x0400: mods.append(keys.?)
+        return mods
+        
+    def _parse_keys(self, e):
+        if e.keysym_num in KEYMAP:
+            return KEYMAP[e.keysym_num], ""
+        # e.char, e.keycode, e.keysym, e.keysym_num
+        key = e.keycode
+        if 97 <= key <= 122:
+            key -= 32
+        if key >= 32 and key <= 127:
+            return keys.Key(chr(key)), chr(key)
+        else:
+            return None, None
+
+    def _on_mouse_enter(self, e):
+        if self._vispy_canvas is None:
+            return
+        self._vispy_mouse_move(
+            pos=(e.x, e.y), modifiers=self._parse_state(e))
+        
+    def _on_mouse_move(self, e):
+        if self._vispy_canvas is None:
+            return
+        self._vispy_mouse_move(
+            pos=(e.x, e.y), modifiers=self._parse_state(e))
+        
+    def _on_mouse_wheel(self, e):
+        if self._vispy_canvas is None:
+            return
+        # e.num: 4 when up, 5 when down
+        # Scroll up: e.delta > 0
+        # Scroll down: e.delta < 0
+        self._vispy_canvas.events.mouse_wheel(
+            delta=(0.0, float(e.delta)), 
+            pos=(e.x, e.y), modifiers=self._parse_state(e))
+        
+    def _on_mouse_button_press(self, e):
+        if self._vispy_canvas is None:
+            return
+        # [ left=1, middle, right]
+        btn = { 1:1, 2:3, 3:2}
+        self._vispy_mouse_press(
+            pos=(e.x, e.y), button=btn[e.num], modifiers=self._parse_state(e))
+        
+    def _on_mouse_double_button_press(self, e):
+        if self._vispy_canvas is None:
+            return
+        # [ left=1, middle, right]
+        btn = { 1:1, 2:3, 3:2}
+        pos, button, modifiers = \
+            (e.x, e.y), btn[e.num], self._parse_state(e)
+        self._vispy_mouse_press(
+            pos=pos, button=button, modifiers=modifiers)
+        self._vispy_mouse_double_click(
+            pos=pos, button=button, modifiers=modifiers)
+
+    def _on_mouse_button_release(self, e):
+        if self._vispy_canvas is None:
+            return
+        # [ left=1, middle, right]
+        btn = { 1:1, 2:3, 3:2}
+        self._vispy_mouse_release(
+            pos=(e.x, e.y), button=btn[e.num], modifiers=self._parse_state(e))
+    
+    def _on_configure(self, e):
+        print("_on_configure")
+        if self._vispy_canvas is None or not self._init:
+            return
+        self.update()
+        self.tkResize(e)
+        self._vispy_canvas.events.resize(size=(e.width, e.height))
+    
+    def _on_draw(self, e):
+        if self._vispy_canvas is None:
+            return
+        if not self._init:
+            self._initialize()
+        self._vispy_canvas.set_current()
+        self._vispy_canvas.events.draw(region=None)
```

Comment:
I noticed that calling `self._vispy_canvas.events.draw(region=None)` just makes the screen black, nothing gets drawn and nothing can be drawn later. For example, when adding `GL.glClear(GL.GL_COLOR_BUFFER_BIT)` to clear the screen (to green on my system), it stays black. Removing the `events.draw` call does make the screen green.

So somehow, it's only clearing the screen to black, but then not draw to it.

Reply:
I think I found it, calling `GL.glClear(GL.GL_COLOR_BUFFER_BIT)` first enables the OpenGL buffer for drawing.